pkg/server/commands/list_objects.go

```diff
@@ -145,34 +158,51 @@ func (q *ListObjectsQuery) evaluate(
 			close(connectedObjectsResChan)
 		}()
 
+		concurrencyLimiterCh := make(chan struct{}, maximumConcurrentChecks)
+
+		wg := sync.WaitGroup{}
+
 		for res := range connectedObjectsResChan {
 
 			if res.ResultStatus == NoFurtherEvalStatus && atomic.AddUint32(objectsFound, 1) <= maxResults {
+				noFurtherEvalRequiredCounter.Inc()
 				resultsChan <- ListObjectsResult{ObjectID: res.Object}
 				continue
 			}
 
-			// todo(optimization): fire this Check resolution off concurrently and continue on to the next object in the
-			// result channel
-			resp, err := q.CheckResolver.ResolveCheck(ctx, &graph.ResolveCheckRequest{
-				StoreID:              req.GetStoreId(),
-				AuthorizationModelID: req.GetAuthorizationModelId(),
-				TupleKey:             tuple.NewTupleKey(res.Object, req.GetRelation(), req.GetUser()),
-				ContextualTuples:     req.GetContextualTuples().GetTupleKeys(),
-				ResolutionMetadata: &graph.ResolutionMetadata{
-					Depth: q.ResolveNodeLimit,
-				},
-			})
-			if err != nil {
-				resultsChan <- ListObjectsResult{Err: err}
-				return // is this the right thing here or should we try to find all of them we can up to some limit?
-			}
-
-			if resp.Allowed && atomic.AddUint32(objectsFound, 1) <= maxResults {
-				resultsChan <- ListObjectsResult{ObjectID: res.Object}
-			}
+			furtherEvalRequiredCounter.Inc()
+
+			wg.Add(1)
+			go func(res *ConnectedObjectsResult) {
+				defer func() {
+					<-concurrencyLimiterCh
```

Comment:
why do we send the signal here instead of entry?
(i.e., should we limit entering this go routine rather than exiting?